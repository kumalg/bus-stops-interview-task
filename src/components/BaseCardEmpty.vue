<template>
    <div ref="cardEl" class="empty-card-container">
        <BaseCard class="empty-card">
            <svg
                class="svg"
                :width="svgSize.width"
                :height="svgSize.height"
                :view-box="`-2 -2 ${svgSize.height} ${svgSize.width}`"
                xmlns="http://www.w3.org/2000/svg"
            >
                <rect
                    class="svg-rect"
                    x="1"
                    y="1"
                    rx="4"
                    ry="4"
                    fill="transparent"
                    :width="cardSize.width"
                    :height="cardSize.height"
                    stroke-width="2px"
                    stroke-dasharray="25 30"
                    stroke-linecap="round"
                />
            </svg>
            <transition name="fade" mode="out-in">
                <p v-if="message" class="message" :key="message">{{ message }}</p>
            </transition>
        </BaseCard>
    </div>
</template>

<script setup lang="ts">
import { computed, onBeforeUnmount, onMounted, ref } from 'vue';

import BaseCard from '@/components/BaseCard.vue';

withDefaults(
    defineProps<{
        message?: string;
    }>(),
    {
        message: ''
    }
);

const cardEl = ref<HTMLDivElement | null>(null);
const cardSize = ref<{ width: number; height: number }>({ width: 0, height: 0 });

const svgSize = computed(() => ({
    width: Math.ceil(cardSize.value.width + 2),
    height: Math.ceil(cardSize.value.height + 2)
}));

const resizeObserver = new ResizeObserver(([entry]) => {
    const { width, height } = entry.contentRect;
    cardSize.value = { width, height };
});

onMounted(() => {
    if (cardEl.value) {
        resizeObserver.observe(cardEl.value);
    }
});

onBeforeUnmount(() => {
    resizeObserver.disconnect();
});
</script>

<style lang="scss" scoped>
.empty-card-container {
    display: flex;
}

.empty-card {
    flex: 1;
    position: relative;
    overflow: initial;

    .svg {
        position: absolute;
        top: -1px;
        left: -1px;
        user-select: none;
        pointer-events: none;
    }

    .svg-rect {
        stroke: $color-gray-600;
    }

    .message {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 100%;
        font-size: 0.875rem;
        line-height: 1.5rem;
        color: $color-gray-700;
        text-align: center;
        margin: 0;
        padding: 0;
    }
}
</style>
