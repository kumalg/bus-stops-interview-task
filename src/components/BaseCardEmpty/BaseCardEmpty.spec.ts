import { flushPromises, mount } from '@vue/test-utils';

import BaseCardEmpty from './BaseCardEmpty.vue';

const EXAMPLE_MESSAGE = 'EXAMPLE_MESSAGE';

describe('BaseCardEmpty', () => {
    const BaseCardEmptyFactory = (props: InstanceType<typeof BaseCardEmpty>['$props'] = {}) => {
        return mount(BaseCardEmpty, {
            props
        });
    };
    let observe: jest.Mock;
    let unobserve: jest.Mock;
    let disconnect: jest.Mock;

    beforeEach(() => {
        observe = jest.fn();
        unobserve = jest.fn();
        disconnect = jest.fn();

        globalThis.ResizeObserver = jest.fn().mockImplementation(() => ({
            observe,
            unobserve,
            disconnect
        }));
    });

    it('renders proper html markup', () => {
        const wrapper = BaseCardEmptyFactory({ message: EXAMPLE_MESSAGE });

        expect(wrapper.element).toMatchSnapshot();
    });

    describe('ResizeObserver', () => {
        it('observe is called after mount', () => {
            BaseCardEmptyFactory({ message: EXAMPLE_MESSAGE });

            expect(observe).toHaveBeenCalledTimes(1);
        });

        it('disconnect is called after removing component from DOM', async () => {
            const wrapper = mount({
                template: `<div><BaseCardEmpty v-if="showCard" /></div>`,
                components: {
                    BaseCardEmpty
                },
                data() {
                    return {
                        showCard: true
                    };
                }
            });

            expect(disconnect).toHaveBeenCalledTimes(0);

            wrapper.setData({ showCard: false });

            await flushPromises();

            expect(disconnect).toHaveBeenCalledTimes(1);
        });
    });
});
