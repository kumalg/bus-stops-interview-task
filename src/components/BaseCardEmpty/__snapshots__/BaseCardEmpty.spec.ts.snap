// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCardEmpty renders proper html markup 1`] = `
<div
  class="empty-card-container"
>
  <div
    class="card empty-card"
  >
    <!--v-if-->
    <div
      class="card-body"
    >
      
      <svg
        class="svg"
        height="2"
        view-box="-2 -2 2 2"
        width="2"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          class="svg-rect"
          fill="transparent"
          height="0"
          rx="4"
          ry="4"
          stroke-dasharray="25 30"
          stroke-linecap="round"
          stroke-width="2px"
          width="0"
          x="1"
          y="1"
        />
      </svg>
      <transition-stub
        appear="false"
        css="true"
        mode="out-in"
        name="fade"
        persisted="false"
      >
        <p
          class="message"
        >
          EXAMPLE_MESSAGE
        </p>
      </transition-stub>
      
    </div>
  </div>
</div>
`;
